File: rust/crates/dirlens-core/src/gitignore.rs
//! .gitignore 内蔵マッチャ（Tier3）。dirlens.py の is_ignored / _extend_pats の等価移植。
//!
//! Tier1（git check-ignore による厳密判定）は session.git_ignored に事前計算した
//! 無視集合が入っている場合に使われる（run.rs で選択）。

use std::path::Path;
use std::sync::Arc;

use crate::cfg::Cfg;
use crate::fnmatch::fnmatch;
use crate::provider::FsProvider;
use crate::session::Session;

/// os.path.relpath 相当（共通接頭辞方式・OS セパレータで連結）。
pub fn relpath(path: &Path, start: &Path) -> String {
    let p: Vec<std::path::Component> = path.components().collect();
    let s: Vec<std::path::Component> = start.components().collect();
    let mut i = 0;
    while i < p.len() && i < s.len() && p[i] == s[i] {
        i += 1;
    }
    let mut parts: Vec<String> = std::iter::repeat("..".to_string())
        .take(s.len() - i)
        .collect();
    for comp in &p[i..] {
        parts.push(comp.as_os_str().to_string_lossy().into_owned());
    }
    if parts.is_empty() {
        ".".to_string()
    } else {
        parts.join(std::path::MAIN_SEPARATOR_STR)
    }
}

/// relpath を "/" 区切りに正規化したもの（Python の .replace("\\", "/") 相当）。
pub fn relpath_slash(path: &Path, start: &Path) -> String {
    relpath(path, start).replace('\\', "/")
}

/// パターンを順番に評価し、最後にマッチしたルールが勝つ（`!` 否定対応）。
pub fn is_ignored(name: &str, rel_path: &str, is_dir: bool, patterns: &[String]) -> bool {
    let rel = rel_path.replace('\\', "/");
    let mut result = false;
    for pat in patterns {
        let negated = pat.starts_with('!');
        let p = pat.trim_start_matches('!');
        let dir_only = p.ends_with('/');
        let p = p.trim_end_matches('/');
        if dir_only && !is_dir {
            continue;
        }
        let matched = if let Some(anchored) = p.strip_prefix('/') {
            let anchored = anchored.trim_start_matches('/');
            fnmatch(&rel, anchored)
        } else {
            fnmatch(name, p) || fnmatch(&rel, p) || fnmatch(&rel, &format!("*/{}", p))
        };
        if matched {
            result = !negated;
        }
    }
    result
}

/// _extend_pats 相当: 下位ディレクトリの .gitignore をルート相対に書き換えて追加する。
pub fn extend_pats<F: FsProvider>(
    sess: &Session<F>,
    active: &Arc<Vec<String>>,
    path: &Path,
    cfg: &Cfg,
) -> Arc<Vec<String>> {
    if !cfg.use_gitignore {
        return active.clone();
    }
    if path == cfg.root.as_path() {
        return active.clone();
    }
    let local = sess.load_gitignore(path);
    if local.is_empty() {
        return active.clone();
    }
    let rel_dir = relpath_slash(path, &cfg.root);
    let mut out: Vec<String> = (**active).clone();
    for pat in local.iter() {
        let neg = pat.starts_with('!');
        let p = pat.trim_start_matches('!');
        if p.starts_with('/') {
            out.push(format!("{}/{}{}", if neg { "!" } else { "" }, rel_dir, p));
        } else {
            out.push(pat.clone());
        }
    }
    Arc::new(out)
}

File: swift/Sources/DirlensCore/Gitignore.swift
// .gitignore 内蔵マッチャ（Tier3）と Tier1 の事前計算
// （rust/crates/dirlens-core/src/gitignore.rs の等価移植）。
//
// Tier1（git check-ignore による厳密判定）は session.gitIgnored に事前計算した
// 無視集合が入っている場合に使われる（Run.swift で選択）。

/// パターンを順番に評価し、最後にマッチしたルールが勝つ（`!` 否定対応）。
public func isIgnored(_ name: String, _ relPath: String, _ isDir: Bool, _ patterns: [String]) -> Bool {
    let rel = relPath
    var result = false
    for pat in patterns {
        let negated = pat.hasPrefix("!")
        var p = Substring(pat)
        while p.hasPrefix("!") { p = p.dropFirst() }
        let dirOnly = p.hasSuffix("/")
        while p.hasSuffix("/") { p = p.dropLast() }
        if dirOnly, !isDir { continue }
        let matched: Bool
        if p.hasPrefix("/") {
            var anchored = p.dropFirst()
            while anchored.hasPrefix("/") { anchored = anchored.dropFirst() }
            matched = fnmatch(rel, String(anchored))
        } else {
            let ps = String(p)
            matched = fnmatch(name, ps) || fnmatch(rel, ps) || fnmatch(rel, "*/\(ps)")
        }
        if matched {
            result = !negated
        }
    }
    return result
}

/// Tier1: git check-ignore による無視集合の事前計算。
///
/// ルートから BFS でレベルごとに全エントリを `git check-ignore --stdin -z` へ
/// 一括投入し、無視された rel パス（"/" 区切り）の集合を作る。無視された
/// ディレクトリには降りない。git が使えない・非 work tree・途中で失敗した場合は
/// nil（Tier3 へ縮退）。
public func buildGitIgnoredSet(_ sess: Session, _ git: GitProvider, _ root: String) -> Set<String>? {
    var ignored: Set<String> = []
    var levelDirs: [String] = [root]
    while !levelDirs.isEmpty {
        var rels: [String] = []
        var children: [FSEntry] = []
        for d in levelDirs {
            if let entries = sess.fs.scanDir(d) {
                for e in entries {
                    rels.append(relpathSlash(e.path, root))
                    children.append(e)
                }
            }
        }
        if rels.isEmpty { break }
        guard let resp = git.checkIgnore(root: root, relPaths: rels) else { return nil }
        ignored.formUnion(resp)
        levelDirs = children
            .filter { $0.isDirNofollow && !ignored.contains(relpathSlash($0.path, root)) }
            .map { $0.path }
    }
    return ignored
}

/// _extend_pats 相当: 下位ディレクトリの .gitignore をルート相対に書き換えて追加する。
public func extendPats(_ sess: Session, _ active: [String], _ path: String, _ cfg: Cfg) -> [String] {
    if !cfg.useGitignore { return active }
    if path == cfg.root { return active }
    let local = sess.loadGitignore(path)
    if local.isEmpty { return active }
    let relDir = relpathSlash(path, cfg.root)
    var out = active
    for pat in local {
        let neg = pat.hasPrefix("!")
        var p = Substring(pat)
        while p.hasPrefix("!") { p = p.dropFirst() }
        if p.hasPrefix("/") {
            out.append("\(neg ? "!" : "")/\(relDir)\(p)")
        } else {
            out.append(pat)
        }
    }
    return out
}
